BinaryTree.java

```diff
@@ -0,0 +1,127 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class BinaryTree<T> {
+    private final BinaryTreeNode<T> root;
+
+    public enum InsertChild {LEFT, RIGHT};
+
+    public BinaryTree(T data) {
+        this.root = new BinaryTreeNode(data);
+    }
+
+    public BinaryTree(BinaryTreeNode<T> root) {
+        this.root = root;
+    }
+
+    public BinaryTreeNode<T> getRoot() {
+        return root;
+    }
+
+    // The function create a left (left or right) to the node with the value parent
+    public void insert(T value, T parent, InsertChild childSide) {
+        // Check if this value already exists in the tree
+        if (this.root.containsKey(value)) {
+            System.err.println("This value already exists");
+            return;
+        }
+
+        // Check if the parent exists in the tree
+        if (!this.root.containsKey(parent)) {
+            System.err.println("This parent isn't exist in the tree");
+            return;
+        }
+        insert(value, this.root.findNode(parent), childSide);
+    }
+
+    // The function create a left (left or right) to the parent node
+    public void insert(T value, BinaryTreeNode<T> parent, InsertChild childSidet) {

```

Comment:
Should probably be private. Only minimal set of methods designed for use from outside should be public. All the methods that expose implementation details (BinaryTreeNodes etc) should be private.